src/ucp/proto/proto_select.c

```diff
@@ -25,6 +28,221 @@ typedef struct {
     size_t                         priv_offsets[UCP_PROTO_MAX_COUNT];
 } ucp_proto_select_init_protocols_t;
 
+/* Temporary list of constructed protocol thresholds */
+typedef struct {
+    size_t                         max_length;
+    ucp_proto_id_t                 proto_id;
+} ucp_proto_threshold_tmp_elem_t;
+
+
+UCS_ARRAY_DEFINE_INLINE(ucp_proto_thresh, unsigned,
+                        ucp_proto_threshold_tmp_elem_t);
+
+const ucp_proto_threshold_elem_t*
+ucp_proto_thresholds_search_slow(const ucp_proto_threshold_elem_t *thresholds,
+                                 size_t msg_length)
+{
+    unsigned idx;
+    for (idx = 0; msg_length > thresholds[idx].max_msg_length; ++idx);
+    return &thresholds[idx];
+}
+
+static ucs_status_t
+ucp_proto_thresholds_append(ucs_array_t(ucp_proto_thresh) *thresh_list,
+                            size_t max_length, ucp_proto_id_t proto_id)
+{
+    ucp_proto_threshold_tmp_elem_t *thresh_elem;
+    ucs_status_t status;
+
+    /* Consolidate with last protocol if possible */
```

Comment:
maybe we should go thru all protocols and try to find the same protocol?
instead of just taking the last protocol and check it

or it is impossible?

Reply:
only consecutive elements which use same protocol can be consolidated, so we can consolidate only with the last protocol